rocks-lib/src/luarocks_installation.rs

```diff
@@ -0,0 +1,221 @@
+pub enum LuaRocksInstallError {
+    #[error(transparent)]
+    Io(#[from] io::Error),
+    #[error(transparent)]
+    ManifestError(#[from] ManifestError),
+    #[error(transparent)]
+    BuildError(#[from] BuildError),
+}
+
+#[derive(Error, Debug)]
+pub enum InstallBuildDependenciesError {
+    #[error(transparent)]
+    Io(#[from] io::Error),
+    #[error(transparent)]
+    ManifestError(#[from] ManifestError),
+    #[error(transparent)]
+    BuildError(#[from] BuildError),
+}
+
+#[derive(Error, Debug)]
+pub enum ExecLuaRocksError {
+    #[error(transparent)]
+    LuaVersionUnset(#[from] LuaVersionUnset),
+    #[error("could not write luarocks config: {0}")]
+    WriteLuarocksConfigError(io::Error),
+    #[error("failed to run luarocks: {0}")]
+    Io(#[from] io::Error),
+    #[error("executing luarocks compatibility layer failed.\nstatus: {status}\nstdout: {stdout}\nstderr: {stderr}")]
+    CommandFailure {
+        status: ExitStatus,
+        stdout: String,
+        stderr: String,
+    },
+}
+
+pub struct LuaRocksInstallation {
+    pub tree: Tree,
+    pub config: Config,
+}
+
+const LUAROCKS_VERSION: &str = "3.11.1-1";
+
+const LUAROCKS_ROCKSPEC: &str = "
+rockspec_format = '3.0'
+package = 'luarocks'
+version = '3.11.1-1'
+source = {
+   url = 'git+https://github.com/luarocks/luarocks',
+   tag = 'v3.11.1'
+}
+";
+
+impl LuaRocksInstallation {
+    pub fn new(config: &Config) -> Result<Self, LuaRocksError> {
+        let config = config.clone().with_tree(config.luarocks_tree().clone());
+        let luarocks_installation = Self {
+            tree: Tree::new(config.luarocks_tree().clone(), LuaVersion::from(&config)?)?,
+            config,
+        };
+        Ok(luarocks_installation)
+    }
+
+    pub async fn ensure_installed(
+        self: &LuaRocksInstallation,
+        progress: &Progress<ProgressBar>,
+    ) -> Result<(), LuaRocksInstallError> {
+        let mut lockfile = self.tree.lockfile()?;
+        let luarocks_req =
+            PackageReq::new("luarocks".into(), Some(LUAROCKS_VERSION.into())).unwrap();
+        if self.tree.has_rock(&luarocks_req).is_none() {
+            let rockspec = Rockspec::new(LUAROCKS_ROCKSPEC).unwrap();
+            let pkg = build(
+                rockspec,
+                PinnedState::Unpinned,
+                LockConstraint::Constrained(luarocks_req.version_req().clone()),
+                BuildBehaviour::NoForce,
+                &self.config,
+                progress,
+            )
+            .await?;
+            lockfile.add(&pkg);
+        }
+        lockfile.flush()?;
+        Ok(())
+    }
+
+    pub async fn install_build_dependencies(
+        self,
+        rockspec: &Rockspec,
+        progress: &Progress<ProgressBar>,
+    ) -> Result<(), InstallBuildDependenciesError> {
+        let mut lockfile = self.tree.lockfile()?;
+        let manifest = ManifestMetadata::from_config(&self.config).await?;
+        let build_dependencies = match rockspec.rockspec_format {
```

Comment:
- [ ] TODO: What if build dependencies themselves have dependencies?

(perhaps for another issue, as this is quite unlikely).